src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_LockStatement.cs

```diff
@@ -37,6 +40,41 @@ public override BoundNode VisitLockStatement(BoundLockStatement node)
                     argumentType); //need to have a non-null type here for TempHelpers.StoreToTemp.
             }
 
+            if (argumentType.IsWellKnownTypeLock() &&
+                TryGetWellKnownTypeMember(lockSyntax, WellKnownMember.System_Threading_Lock__EnterLockScope, out MethodSymbol enterLockScope) &&
+                TryGetWellKnownTypeMember(lockSyntax, WellKnownMember.System_Threading_Lock__Scope__Dispose, out MethodSymbol lockScopeDispose))
+            {
+                // lock (x) { body } -> using (x.EnterLockScope()) { body }
+
+                var tryBlock = rewrittenBody is BoundBlock block ? block : BoundBlock.SynthesizedNoLocals(lockSyntax, rewrittenBody);
+
+                var enterLockScopeCall = BoundCall.Synthesized(
+                    rewrittenArgument.Syntax,
+                    rewrittenArgument,
+                    initialBindingReceiverIsSubjectToCloning: ThreeState.Unknown,
+                    enterLockScope);
+
+                BoundLocal boundTemp = _factory.StoreToTemp(enterLockScopeCall,
+                    out BoundAssignmentOperator tempAssignment,
+                    syntaxOpt: rewrittenArgument.Syntax,
+                    kind: SynthesizedLocalKind.Using);
+                var expressionStatement = new BoundExpressionStatement(rewrittenArgument.Syntax, tempAssignment);
+
+                BoundStatement tryFinally = RewriteUsingStatementTryFinally(
+                    rewrittenArgument.Syntax,
+                    rewrittenArgument.Syntax,
+                    tryBlock,
+                    boundTemp,
+                    awaitKeywordOpt: default,
+                    awaitOpt: null,
+                    patternDisposeInfo: MethodArgumentInfo.CreateParameterlessMethod(lockScopeDispose));
+
+                return new BoundBlock(
+                    lockSyntax,
+                    locals: ImmutableArray.Create(boundTemp.LocalSymbol),
+                    statements: ImmutableArray.Create(expressionStatement, tryFinally));
+            }
+
             if (argumentType.Kind == SymbolKind.TypeParameter)
```

Comment:
>if (argumentType.Kind == SymbolKind.TypeParameter) [](http://example.com/codeflow?start=12&length=50)

The fact that we can get here for a Lock type is surprising to me.

Reply:
> The fact that we can get here for a Lock type is surprising to me.

This thread was marked as resolved, but I do not see a relevant change in code.  This point is still reachable when ```LockBinder.TryFindLockTypeInfo``` fails. If we do not expect it to fail, then we could simply call it inside the previous `if` (remove it from the condition) and assert that it succeeds, or restructure that code some other way to make this statement unreachable when ```argumentType.IsWellKnownTypeLock()``` is true.

Reply:
> The fact that we can get here for a Lock type is surprising to me.

This looks resolved in commit 6